server/application/websocket_test.go

```diff
@@ -164,3 +165,147 @@ func TestValidateWithoutPermissions(t *testing.T) {
 
 	testServerConnection(t, validate, true)
 }
+
+func TestTerminalSession_Read(t *testing.T) {
+	validate := func(w http.ResponseWriter, r *http.Request) {
+		ts := newTestTerminalSession(w, r)
+		if ts.wsConn == nil {
+			t.Fatalf("WebSocket connection is not initialized")
+		}
+
+		ts.terminalOpts = &TerminalOptions{DisableAuth: true}
+
+		testMessages := []TerminalMessage{
+			{
+				Operation: "stdin",
+				Data:      "test input",
+			},
+			{
+				Operation: "resize",
+				Cols:      80,
+				Rows:      24,
+			},
+			{
+				Operation: "unknown",
+			},
+		}
+
+		for _, msg := range testMessages {
+			bytes, _ := json.Marshal(msg)
+			err := ts.wsConn.WriteMessage(websocket.TextMessage, bytes)
+			if err != nil {
+				t.Errorf("Failed to write message: %v", err)
+			}
+		}
+	}
+
+	tests := []struct {
+		name           string
+		expectedOutput string
+		expectedSize   *remotecommand.TerminalSize
+		expectedError  bool
+	}{
+		{
+			name:           "stdin operation",
+			expectedOutput: "test input",
+			expectedSize:   nil,
+			expectedError:  false,
+		},
+		{
+			name:           "resize operation",
+			expectedOutput: "", // Expecting no output for resize
+			expectedSize: &remotecommand.TerminalSize{
+				Width:  80,
+				Height: 24,
+			},
+			expectedError: false,
+		},
+		{
+			name:           "unknown operation",
+			expectedOutput: EndOfTransmission,
+			expectedSize:   nil,
+			expectedError:  true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			s := httptest.NewServer(http.HandlerFunc(validate))
+			defer s.Close()
+
+			u := "ws" + strings.TrimPrefix(s.URL, "http")
+			ws, _, err := websocket.DefaultDialer.Dial(u, nil)
+			require.NoError(t, err)
+			defer ws.Close()
+
+			ts := terminalSession{
+				wsConn:       ws,
+				sizeChan:     make(chan remotecommand.TerminalSize, 1),
+				terminalOpts: &TerminalOptions{DisableAuth: true},                                                            // Ensure terminalOpts is initialized
+				token:        new(string),                                                                                    // Initialize token to avoid nil dereference
+				ctx:          context.WithValue(context.Background(), "claims", &jwt.MapClaims{"groups": []string{"admin"}}), // Set context with claims
+			}
+
+			p := make([]byte, 1024)
+			n, err := ts.Read(p)
+
+			if tt.expectedError {
+				assert.Error(t, err)
+				assert.Equal(t, tt.expectedOutput, string(p[:n]))
+			} else {
+				assert.NoError(t, err)
+				assert.Equal(t, tt.expectedOutput, string(p[:n]))
+			}
+
+			if tt.expectedSize != nil {
+				select {
+				case size := <-ts.sizeChan:
+					assert.Equal(t, *tt.expectedSize, size)
+				default:
+					t.Error("Expected size message not received")
+				}
+			}
+		})
+	}
+}
+
+func TestTerminalSession_Write(t *testing.T) {
```

Comment:
Just a personal tip but in general I think it is preferable to have a more verbose description of what exactly you are validating as part of your unit test. See this test for reference:
https://github.com/argoproj/argo-cd/blob/00466c30949c2d85bda3073d97049363e21333c5/util/argo/managedfields/managed_fields_test.go#L21

Reply:
I have changed the code structure this time, can you please take a look?